### PR TITLE
Add extensions as init container

### LINKED ISSUE #9785: Extensions support in the operator
* Add a new field to Keycloak CRD for specifying a list of URLs to extensions.
* The extensions will be downloaded and configured to be used in Keycloak.

Requires investigation on how this can work with the seamless augmentation.

See also:
* #9605
* #9437

### PR DESCRIPTION
**Pull request by andreaTP**
This is a stop-gap solution for #9785 , I open it in draft since we need to discuss more.

Resolves: #9785

### CODE REVIEW

**Review by pedroigor — DISMISSED** [review]

**Review by vmuzikar — CHANGES_REQUESTED** [review]
Thanks @andreaTP, added a few comments.

Before merging this, we need to align with @stianst on the container that will be used. In the operator code, we're [using](https://github.com/keycloak/keycloak/blob/26df1b007d8ac66796961af2d17c2f59914949c0/operator/src/main/resources/application.properties#L11) [existing init container](https://quay.io/repository/keycloak/keycloak-init-container) what was originally meant for WF dist. But in the CI, we're [using](https://github.com/keycloak/keycloak/blob/26df1b007d8ac66796961af2d17c2f59914949c0/.github/workflows/operator-ci.yml#L53-L57) a [new init container](https://github.com/keycloak/keycloak/tree/26df1b007d8ac66796961af2d17c2f59914949c0/quarkus/init-container). This needs to be aligned. I see 3-4 options.
1. Rely only on the [`legacy` tag](https://quay.io/repository/keycloak/keycloak-init-container?tab=tags) (effectively it is latest) of init container both in operator and the tests. We don't plan any changes in it anyway, so it should be ok-ish. It is an stop-gap solution in any case, before we switch to some `kc.sh` command.
2. Rely only on the `legacy` tag but checkout the [keycloak-containers repo](https://github.com/keycloak/keycloak-containers) and build it in the CI for tests. It's not ideal but we'd still catch potential issues sooner.
3. Rely on the[ new init container](https://github.com/keycloak/keycloak/tree/26df1b007d8ac66796961af2d17c2f59914949c0/quarkus/init-container) both in operator and the CI.
4. Per @stianst suggestion, we could go ahead and really implement it as a simple `kc.sh extensions` command without using an init container at all.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java [hunk 1]
```diff
@@ -30,10 +33,14 @@
 import org.keycloak.operator.v2alpha1.crds.KeycloakStatusBuilder;
 
 import java.net.URL;
+import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Optional;
 import java.util.stream.Collectors;
 
+import static org.keycloak.operator.Constants.*;
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:42]: done

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java [hunk 2]
```diff
@@ -87,6 +94,63 @@ private Deployment fetchExistingDeployment() {
...
+                .withName(EXTENSIONS_VOLUME_NAME)
+                .withMountPath(KEYCLOAK_PROVIDERS_FOLDER)
+                .withReadOnly(true)
+                .build();
+        containerVolumeMounts.add(extensionVM);
+
+        container.setVolumeMounts(containerVolumeMounts);
+
+        // Add the Extensions downloader init container
+        var initContainers = baseDeployment.getSpec().getTemplate().getSpec().getInitContainers();
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:133]: removed

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java [hunk 3]
```diff
@@ -87,6 +94,63 @@ private Deployment fetchExistingDeployment() {
...
+                .withMountPath(INIT_CONTAINER_EXTENSIONS_FOLDER)
+                .endVolumeMount()
+                .addNewEnv()
+                .withName(INIT_CONTAINER_EXTENSIONS_ENV_VAR)
+                .withValue(extensionsValue)
+                .endEnv()
+                .build();
+
+        initContainers.add(initContainer);
+        baseDeployment.getSpec().getTemplate().getSpec().setInitContainers(initContainers);
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:151]: I think this is good practice ... but not fussy, fixed

File: operator/src/main/resources/application.properties [hunk 4]
```diff
@@ -5,5 +5,8 @@ quarkus.container-image.builder=jib
 quarkus.operator-sdk.crd.validate=false
 
 # Operator config
-operator.keycloak.image=quay.io/keycloak/keycloak-x:latest
+operator.keycloak.image=quay.io/keycloak/keycloak:latest
 operator.keycloak.image-pull-policy=Always
+
+operator.keycloak.init-container-image=quay.io/keycloak/keycloak-init-container:latest
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/main/resources/application.properties:11]: done

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 5]
```diff
@@ -113,6 +116,36 @@ public void testDeploymentDurability() {
         }
     }
 
+    @Test
+    public void testExtensions() {
+        try {
+            var kc = getDefaultKeycloakDeployment();
+            kc.getSpec().setExtensions(
+                    Collections.singletonList(
+                            "https://github.com/aerogear/keycloak-metrics-spi/releases/download/2.5.3/keycloak-metrics-spi-2.5.3.jar"));
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:125]: make sense 👍

**Comment by andreaTP** [comment]
@vmuzikar thanks for the review, I believe I addressed all the outstanding comments.

Regarding alignment on which init container to use, I think that the best option is to use the new init container(3) in operator and CI. Happy to complete this PR automating the new container publishing (even if I would need some help for doing it :-) ).

> implement it as a simple `kc.sh extensions` command without using an init container at all.

We already discussed and agreed, as a team, that this is not the right time to go in this direction(keeping the door open for it in the future).

**Review by abstractj — COMMENTED** [review]
@andreaTP could you please rebase?

**Comment by andreaTP** [comment]
@abstractj done, ready for review again 🙂

**Comment by andreaTP** [comment]
@vmuzikar this should be ready for a final review.

**Review by vmuzikar — APPROVED** [review]
LGTM, thanks @andreaTP! This is ready to be merged.

**Review by pedroigor — APPROVED** [review]

**Pull request merged**